### ISSUE TITLE
Parallel setting not working

### ISSUE THREAD

**Issue by marboleda** [javierbrea/cypress-fail-fast#244]
**Describe the bug**
I would like to use the setting to skip all tests in parallel runs. But I am getting the following error after my failed test:

```
"after each" hook for "<Test Title>":
     CypressError: `cy.task('failFastShouldSkip')` failed with the following error:
> The "data" argument must be of type string or an instance of Buffer, TypedArray, or DataView. Received undefined
https://on.cypress.io/api/task
Because this error occurred during a `after each` hook we are skipping all of the remaining tests.
```

None of the specs afterwards or the ones in the parallel jobs skip after this failure, they just run as usual

**To Reproduce**
This is happening in a GitLab CI job. This is the command used to run the tests:

`cypress run-ct --spec <spec_file_1>, <spec_file_2>, ...`

I set the configuration for parallel runs the same as the [example in the README](https://github.com/javierbrea/cypress-fail-fast#configuration-for-parallel-runs).

The "run" mode works as expected (skips remaining tests within a job if one fails), it's just with "parallel" that I am experiencing this.

**Expected behavior**
Remaining tests in job and the ones running in parallel should skip as soon as this test fails

**Operating system, Node.js an npm versions, or browser version (please complete the following information):**
 - Node.js: v16.14.2
 - Cypress: 9.7.0
 - Browser: Electron 100 (headless)

**Comment by javierbrea** [comment]
Hi @marboleda ,

I suspect that it may be related to an error produced in one of your `parallelCallbacks`. After taking a look at the code, if an error is produced in the callback, the `failFastShouldSkip` task would return undefined, and this produces a Cypress error.

Could you check if the `.run-is-cancelled` file really exists in the project folder after a failing test? Or is your callback throwing an error because it couldn't write the file? You could also add `console.log` statements to your callbacks for debugging the problem. Those logs would be printed in Cypress logs, so you can easily check them.

Note also that the provided example is only for demonstration purposes. In order to it to work, all processes should have shared access to the folder in which the "flag" file is written, as it is mentioned in the docs. Otherwise all of the other processes would continue running, because they wouldn't be able to know if another parallel callback failed.

Anyway, the plugin could be improved by catching any possible error produced in the parallel callbacks, and, at least, inform the user about it. So, I'll open an issue for that.

Thanks for the feedback, and please keep me up to date about the problem!

**Comment by marboleda** [comment]
Hi @javierbrea,

Thank you for the response, I drilled down into the error messages further and found that there were references to fs.writeFileSync.

It turns out that I needed to provide a second argument as it is [not optional](https://nodejs.org/api/fs.html#fswritefilesyncfile-data-options). This is the "data" argument being referenced in the error message. Supplying it with an empty string removes the error like so:

`fs.writeFileSync(isCancelledFlagFile, '');`

The rest of the tests in the job skip properly in parallel mode after making this change.

This still doesn't create the situation I was trying to accomplish, as I was attempting to stop tests that are running in jobs that are running in parallel at the same time. However, after doing [further research](https://stackoverflow.com/a/69669708), it may not really be possible with this approach. The only way GitLab can pass data from one job to another is through artifacts and those do not get exposed until a job is completed.

There may be another solution involving writing to a file somewhere else and reading that, but we can decide later on if that would be worth the effort.

Other than modfying the example in the README (and perhaps the error logging improvement you mentioned), no other work needs to be done on this issue. Thank you for your assistance!

**Comment by javierbrea** [comment]
Hi again @marboleda ,

I have opened issues #246 and #245 for the mentioned problems.

About possible alternatives to use a file as a "flag" for stopping parallel runs, I can tell you about a solution that I was about to implement in one of my latest projects. Maybe it can be also valid for you, but I warn you that it requires extra "infrastructure", so very probably it neither does worth.

The designed solution was to implement a very simple external REST API with next endpoints:

* POST /test-executions/ - Stores an item in DDBB with a `status` property, creating an id automatically. Returns the item.
* PATCH /test-executions/ - Modifies the item.
* GET /test-executions/{id} - Returns an item by id.

Then, just in a step before starting any of the parallel runs, the pipeline should POST a new "test-execution" with status "running". It should get the generated item ID in order to pass it to all of the Cypress parallel processes as an environment variable.

In the `isCancelled` callback, it should `GET /test-executions/{id}` and check if the status is equal to "running". If it isn't, the callback must return `true`. In the `onCancel` parallel callback, it should send a `PATCH` to `/test-executions/{id}` updating the status to "failed".

So, apart from providing parallel fail fast, the API could also serve as some sort of tests execution dashboard which could be improved afterwards updating the result to "success" when all tests passed, adding start and end timestamps, etc.

...and, now that I'm reviewing this design and the Cypress docs, I have just noticed that Cypress node events support asynchronies, while the plugin callbacks don't. I don't know if it is only supported recently and it wasn't when I created the plugin, or simply I forgot this detail 😓, but I will open another issue for the plugin in order to add support for parallel callbacks returning promises. 3 issues opened from this one, not bad! 😉 Thanks again for the feedback!
